### PR TITLE
[next] 1358 - MainPage (index.js) 

### LINKED ISSUE #1358: [next] Main page, see src/frontend/src/pages/index.js
**What would you like to be added**:
[Next.js migration. Port-MAINPAGE](https://github.com/Seneca-CDOT/telescope/issues/1316)
**Why would you like this to be added**:
Next.js migration.

### PR DESCRIPTION
**Pull request by PedroFonsecaDEV**
## Issue This PR Addresses

Fixes #1358 - [next] Port MainPage.

## Type of Change
PORT TO NEXT.JS 

- [ ] **Bugfix**: Change which fixes an issue
- [x] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description
The MainPage.tsx is done; I had to create some basic components to have it. 

It seems to me that we don't need the PageBase component now. 

## Checklist

- [ ] **Quality**: This PR builds and passes our npm test and works locally
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

**Comment by humphd** [comment]
Just starting to look at this, and wondering if you have tips on how to test this?  @PedroFonsecaDEV

File: src/frontend/next/src/pages/index.tsx [hunk 1]
```diff
@@ -1,60 +1,25 @@
+import { FC } from 'react';
 import Head from 'next/head';
-import styles from '../styles/Home.module.css';
```
Comment (humphd) [inline src/frontend/next/src/pages/index.tsx:2]: It looks like we haven't got CSS figured out yet, is that happening in a follow-up?

File: src/frontend/next/src/pages/index.tsx [hunk 2]
```diff
@@ -1,60 +1,25 @@
...
 
-const Home = () => {
+import Banner from '../components/Banner/Banner';
+import BackToTopButton from '../components/BackToTopButton/BackToTopButton';
+import Posts from '../components/Posts/Posts';
+
+const MainPage: FC = () => {
   return (
-    <div className={styles.container}>
+    <>
```
[the remark this answers is not in the document]
Comment (HyperTHD) [reply src/frontend/next/src/pages/index.tsx:10]: It's a React Fragment, which works in a similar area. refer [here](https://stackoverflow.com/questions/47761894/why-are-fragments-in-react-16-better-than-container-divs) for it's possible uses

**Comment by PedroFonsecaDEV** [comment]
@humphd 
> Just starting to look at this, and wondering if you have tips on how to test this? @PedroFonsecaDEV

I use the Netlify deployment to test it here on GitHub.

File: src/frontend/next/src/pages/index.tsx (hunk 1 above)
Comment (PedroFonsecaDEV) [reply src/frontend/next/src/pages/index.tsx:2]: File: [telescope/src/frontend/gatsby/src/pages/index.js](https://github.com/Seneca-CDOT/telescope/blob/master/src/frontend/gatsby/src/pages/index.js)

```js
import React from 'react';

import PageBase from './PageBase';
import Banner from '../components/Banner';
import Posts from '../components/Posts';
import BackToTopButton from '../components/BackToTopButton';

export default function IndexPage() {
  return (
    <PageBase title="Home">
      <Banner />
      <BackToTopButton />
      <main className="main">
        <Posts />
      </main>
    </PageBase>
  );
}
```

We don't have a CSS in this component. The `PageBase` component holds the global `CSS` and `SEO`. 
With Next.js, we may have to break the `PageBase`. Reading this [Next.js doc](https://nextjs.org/docs/basic-features/built-in-css-support) I understood that the "global CSS" would be defined on the file: [next/src/pages/_app.tsx](https://github.com/Seneca-CDOT/telescope/blob/master/src/frontend/next/src/pages/_app.tsx). But I would appreciate feedback on this approach.  

Note about `SEO`: I'm not sure if we could add our `SEO` to `_app.tsx` or if we need to add it to each page.

File: src/frontend/next/src/pages/index.tsx [hunk 3]
```diff
@@ -1,60 +1,25 @@
+import { FC } from 'react';
 import Head from 'next/head';
-import styles from '../styles/Home.module.css';
 
-const Home = () => {
+import Banner from '../components/Banner/Banner';
+import BackToTopButton from '../components/BackToTopButton/BackToTopButton';
+import Posts from '../components/Posts/Posts';
+
+const MainPage: FC = () => {
```
[the remark this answers is not in the document]
Comment (PedroFonsecaDEV) [reply src/frontend/next/src/pages/index.tsx:8]: Sure.
In our Gatsby we export this [component](https://github.com/Seneca-CDOT/telescope/blob/master/src/frontend/gatsby/src/pages/index.js#L8) as `IndexPage`. 
and
In our [Port](https://github.com/Seneca-CDOT/telescope/issues/1316) issue we refer to this component as `Main Page`.

I will change it to `Home`.

File: src/frontend/next/src/pages/index.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (PedroFonsecaDEV) [reply src/frontend/next/src/pages/index.tsx:10]: [React Fragments docs](https://reactjs.org/docs/fragments.html)

File: src/frontend/next/src/pages/index.tsx (hunk 1 above)
Comment (tonyvugithub) [reply src/frontend/next/src/pages/index.tsx:2]: @PedroFonsecaDEV : We can defined global CSS in MUI theme as well. I think that would be a better approach to have global styling in the theme.

**Comment by tonyvugithub** [comment]
I am still leaning toward keeping the PageBase component for quick migration. The least adjustment we need to make the better. How you guys think?

File: src/frontend/next/src/pages/index.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (chrispinkney) [reply src/frontend/next/src/pages/index.tsx:10]: Thanks guys!

File: src/frontend/next/src/pages/index.tsx (hunk 1 above)
Comment (humphd) [reply src/frontend/next/src/pages/index.tsx:2]: For global styling, the work happening in #1549 is probably the right approach.

**Comment by PedroFonsecaDEV** [comment]
@tonyvugithub 
> I am still leaning toward keeping the PageBase component for quick migration. The least adjustment we need to make the better. How you guys think?

I'm working on the `pageBase` problem (https://github.com/Seneca-CDOT/telescope/issues/1548). Probably I will break It into small PRs. And the first one is gonna be the `pageBase` solution.

File: src/frontend/next/src/components/BackToTopButton/BackToTopButton.tsx [hunk 4]
```diff
@@ -0,0 +1,5 @@
+import { FC } from 'react';
+
+const BackToTopButton: FC = () => <h3>BackToTopButton Component</h3>;
+
+export default BackToTopButton;
```
Comment (chrispinkney) [inline src/frontend/next/src/components/BackToTopButton/BackToTopButton.tsx:5]: In a few of our current PRs we're adding an `index.ts` file for easier imports, [see here](https://github.com/Seneca-CDOT/telescope/pull/1513/files#diff-23afde8fc396f80b6fe6e1c7acf70eebf18273d03d482bc0266ae5671cadda4cR1) for an example, can you convert `BackToTopButton`, `Banner`, and `Posts` into that? The imports will need to get changed after too 😄

**Comment by tonyvugithub** [comment]
I wonder if we can merge this now or wait after we merge BackTopTopBUtton, Header and Posts to merge it. What would be a better approach here. This looks good to go though

File: src/frontend/next/src/components/BackToTopButton/BackToTopButton.tsx (hunk 4 above)
Comment (humphd) [reply src/frontend/next/src/components/BackToTopButton/BackToTopButton.tsx:5]: This is a great point to raise, @chrispinkney.  Here are my thoughts:

- JSX (and for TypeScript, TSX) is a non-standard extension of JavaScript
- Some people, especially outside the React ecosystem, don't like using `.js` for a file that isn't standard JavaScript (or `.ts` for TypeScript). 
- Transpilers (i.e., Babel, which translates JSX to pure JS) doesn't care what you call it, and at runtime, there is no-such thing as JSX: it all gets converted to pure functions.
- The general advice with transpiled code is to import it **without** the extension: `import Foo from './foo'` without any extension, and let the module system figure it out at build- vs. run-time.
- We are using React and next.js, so the vast majority of files are likely to contain JSX (i.e., it's not an edge case we need to worry about).

I think that another reason we used the `index.js` approach to do the final export was that we also wanted to put CSS files in the same scope:

```
/Foo
    index.js    <---- takes care of combining the JSX and CSS, and provides the default export for Foo
    Foo.jsx
    Foo.css
```

Later, we moved to Material UI, and with it, CSS-in-JS.  Next.js doesn't really like global CSS anyway (they either want you to use CSS Modules or CSS-in-JS).

In summary, I think we can ditch the `index.js` export files completely.  If we follow the patterns in the [next.js TypeScript example](https://github.com/vercel/next.js/tree/canary/examples/with-typescript), in this PR I would suggest we do this:

`src/frontend/next/src/components/BackToTopButton/BackToTopButton.tsx` becomes ` src/frontend/next/src/components/BackToTopButton.tsx` (i.e., drop the directory and use `tsx`).

**Review by humphd — CHANGES_REQUESTED** [review]

File: src/frontend/next/src/components/BackToTopButton/BackToTopButton.tsx (hunk 4 above)
Comment (humphd) [reply src/frontend/next/src/components/BackToTopButton/BackToTopButton.tsx:5]: Filed https://github.com/Seneca-CDOT/telescope/issues/1553 for follow-up.

Comment (humphd) [reply src/frontend/next/src/components/BackToTopButton/BackToTopButton.tsx:5]: Filed https://github.com/Seneca-CDOT/telescope/issues/1554 for follow-up.

**Review by humphd — APPROVED** [review]
I think we can merge this, and then replace things when the other components are done.  It would be good to get a main page that works.

**Review by manekenpix — APPROVED** [review]
LGTM :+1: 

image
